qurator/dinglehopper/extracted_text.py

```diff
@@ -157,6 +179,35 @@ def text(self):
         else:
             return self.joiner.join(s.text for s in self.segments)
 
+    @functools.cached_property
+    def _joiner_grapheme_cluster(self):
+        """We need the joiner as a list of 0 or 1 grapheme clusters.
+
+        This property is cached.
+        """
+
+        if len(self.joiner) > 0:
+            joiner_grapheme_cluster = list(grapheme_clusters(self.joiner))
+            assert len(joiner_grapheme_cluster) == 1  # see joiner's check above
+        elif len(self.joiner) == 0:
+            joiner_grapheme_cluster = []
+        else:
+            joiner_grapheme_cluster = None
+
+        return joiner_grapheme_cluster
+
+    @property
+    def grapheme_clusters(self):
+        if self._text is not None:
+            return self._grapheme_clusters
+        else:
+            # TODO Test with text extracted at glyph level (joiner == "")
+            clusters = []
+            for seg in self.segments:
+                clusters += seg.grapheme_clusters + self._joiner_grapheme_cluster
+            clusters = clusters[:-1]
+            return clusters
```

Comment:
The previous change assumed that the joiner is exactly one grapheme cluster. Well, most of the time - it can also be 0 grapheme clusters, in case we extracted at glyph level (means we glue single grapheme clusters together with nothing). Could be OK now but I need to add a test for this edge case. (Don't think anyone uses it anyway.)

Reply:
Left a TODO re: test in the code.